src/main/java/de/tum/cit/aet/artemis/core/service/course/CourseAccessService.java

```diff
@@ -0,0 +1,232 @@
+     * @param course The course to which the user should get added to
+     */
+    public void enrollUserForCourseOrThrow(User user, Course course) {
+        authCheckService.checkUserAllowedToEnrollInCourseElseThrow(user, course);
+        userService.addUserToGroup(user, course.getStudentGroupName());
+        if (course.getLearningPathsEnabled()) {
+            learnerProfileApi.ifPresent(api -> api.createCourseLearnerProfile(course, user));
+            learningPathApi.ifPresent(api -> api.generateLearningPathForUser(course, user));
+        }
+        final var auditEvent = new AuditEvent(user.getLogin(), Constants.ENROLL_IN_COURSE, "course=" + course.getTitle());
+        auditEventRepository.add(auditEvent);
+        log.info("User {} has successfully enrolled in course {}", user.getLogin(), course.getTitle());
+    }
+
+    /**
+     * Add multiple users to the course so that they can access it
+     * The passed list of UserDTOs must include at least one unique user identifier (i.e. registration number OR email OR login)
+     * <p>
+     * This method first tries to find the user in the internal Artemis user database (because the user is probably already using Artemis).
+     * In case the user cannot be found, it additionally searches the connected LDAP in case it is configured.
+     *
+     * @param courseId    the id of the course
+     * @param studentDTOs the list of students (with at least registration number)
+     * @param courseGroup the group the students should be added to
+     * @return the list of students who could not be enrolled in the course, because they could NOT be found in the Artemis database and could NOT be found in the TUM LDAP
+     */
+    public List<StudentDTO> registerUsersForCourseGroup(Long courseId, List<StudentDTO> studentDTOs, String courseGroup) {
+        var course = courseRepository.findByIdElseThrow(courseId);
+        if (course.getLearningPathsEnabled()) {
+            course = courseRepository.findWithEagerCompetenciesAndPrerequisitesByIdElseThrow(course.getId());
+        }
+        String courseGroupName = course.defineCourseGroupName(courseGroup);
+        Role courseGroupRole = Role.fromString(courseGroup);
+        List<StudentDTO> notFoundStudentsDTOs = new ArrayList<>();
+        for (var studentDto : studentDTOs) {
+            var optionalStudent = userService.findUserAndAddToCourse(studentDto.registrationNumber(), studentDto.login(), studentDto.email(), courseGroupName);
+            if (optionalStudent.isEmpty()) {
+                notFoundStudentsDTOs.add(studentDto);
+            }
+            else if (courseGroupRole == Role.STUDENT && course.getLearningPathsEnabled()) {
+                final Course finalCourse = course;
+                learnerProfileApi.ifPresent(api -> api.createCourseLearnerProfile(finalCourse, optionalStudent.get()));
+                learningPathApi.ifPresent(api -> api.generateLearningPathForUser(finalCourse, optionalStudent.get()));
+            }
```

Comment:
No strong opinion here, I guess that is opinionated
```suggestion
            Optional<User> student = userService.findUserAndAddToCourse(studentDto.registrationNumber(), studentDto.login(), studentDto.email(), courseGroupName);
            if (student.isEmpty()) {
                notFoundStudentsDTOs.add(studentDto);
            }
            else if (courseGroupRole == Role.STUDENT && course.getLearningPathsEnabled()) {
                final Course finalCourse = course;
                learnerProfileApi.ifPresent(api -> api.createCourseLearnerProfile(finalCourse, student.get()));
                learningPathApi.ifPresent(api -> api.generateLearningPathForUser(finalCourse, student.get()));
            }
```